lib/vnet/vnet_test.go

```diff
@@ -0,0 +1,281 @@
+	}
+
+	utils.RunTestBackgroundTask(ctx, t, &utils.TestBackgroundTask{
+		Name: "test network stack",
+		Task: func(ctx context.Context) error {
+			if err := forwardBetweenTunAndNetstack(ctx, tun1, linkEndpoint); !errIsOK(err) {
+				return trace.Wrap(err)
+			}
+			return nil
+		},
+		Terminate: func() error {
+			linkEndpoint.Close()
+			return trace.Wrap(tun1.Close())
+		},
+	})
+
+	// Assign a local IP address to the test stack.
+	localAddress, err := randomULAAddress()
+	require.NoError(t, err)
+	protocolAddr, err := protocolAddress(localAddress)
+	require.NoError(t, err)
+	tcpErr := testStack.AddProtocolAddress(nicID, protocolAddr, stack.AddressProperties{})
+	require.Nil(t, tcpErr)
+
+	// Route the VNet range to the TUN interface - this emulates the route that will be installed on the host.
+	vnetIPv6Prefix, err := IPv6Prefix()
+	require.NoError(t, err)
+	subnet, err := tcpip.NewSubnet(vnetIPv6Prefix, tcpip.MaskFromBytes(net.CIDRMask(64, 128)))
+	require.NoError(t, err)
+	testStack.SetRouteTable([]tcpip.Route{{
+		Destination: subnet,
+		NIC:         nicID,
+	}})
+
+	// Create the VNet and connect it to the other side of the TUN.
+	manager, err := NewManager(&Config{
+		TUNDevice:  tun2,
+		IPv6Prefix: vnetIPv6Prefix,
+	})
+	require.NoError(t, err)
+
+	utils.RunTestBackgroundTask(ctx, t, &utils.TestBackgroundTask{
+		Name: "VNet",
+		Task: func(ctx context.Context) error {
+			if err := manager.Run(ctx); !errIsOK(err) {
+				return trace.Wrap(err)
+			}
+			return nil
+		},
+		Terminate: func() error {
+			tunErr := tun2.Close()
+			destroyErr := manager.Destroy()
+			return trace.NewAggregate(tunErr, destroyErr)
+		},
+	})
+
+	return &testPack{
+		vnetIPv6Prefix:   vnetIPv6Prefix,
+		manager:          manager,
+		testStack:        testStack,
+		testLinkEndpoint: linkEndpoint,
+		localAddress:     localAddress,
+	}
+}
+
+// dial dials the VNet address [addr] from the test virtual netstack.
+func (p *testPack) dial(ctx context.Context, addr tcpip.Address) (*gonet.TCPConn, error) {
+	conn, err := gonet.DialTCPWithBind(
+		ctx,
+		p.testStack,
+		tcpip.FullAddress{
+			NIC:      nicID,
+			Addr:     p.localAddress,
+			LinkAddr: p.testLinkEndpoint.LinkAddress(),
+		},
+		tcpip.FullAddress{
+			NIC:      nicID,
+			Addr:     addr,
+			Port:     456,
+			LinkAddr: p.manager.linkEndpoint.LinkAddress(),
+		},
+		ipv6.ProtocolNumber,
+	)
+	return conn, trace.Wrap(err)
+}
+
+// TestVnetEcho is a preliminary VNet test that manually installs an echo handler on a specific IP, TCP dials
+// it, and makes sure writes are echoed back to the TCP conn.
+func TestVnetEcho(t *testing.T) {
+	ctx := context.Background()
+	p := newTestPack(t, ctx)
+
+	dialAddress, err := p.manager.assignTCPHandler(echoHandler{})
+	require.NoError(t, err)
+
+	for i := 0; i < 10; i++ {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			t.Parallel()
+
+			conn, err := p.dial(ctx, dialAddress)
+			require.NoError(t, err)
+			defer conn.Close()
+
+			testString := "Hello, World!\n"
```

Comment:
It seems like this `\n` is load bearing, but why? (Maybe worth a comment).

Reply:
so the bufio scanner can read a line, updated the comment. this test is mostly being replaced later otherwise i'd remove the bufio